### PR TITLE
Allow Bearer JWT client authentication

### PR DESCRIPTION
**Pull request by amydutta**
Allow Bearer JWT client authentication
Added tests
Re-formatted

### CODE REVIEW

**Comment by boumenot** [comment]
Does this change require documentation?
I read your commit message, and it does not appear to do justice to the amount of change in this PR.

File: builder/azure/arm/clientconfig.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/Azure/go-autorest/autorest/azure"
+	jwt "github.com/dgrijalva/jwt-go"
+	packerAzureCommon "github.com/hashicorp/packer/builder/azure/common"
+	"github.com/hashicorp/packer/packer"
+)
+
+// ClientConfig allows for various ways to authenticate Azure clients
+type ClientConfig struct {
+	// Describes where API's are
```
Comment (boumenot) [inline builder/azure/arm/clientconfig.go]: Where's the description?

Comment (paulmey) [reply builder/azure/arm/clientconfig.go]: in the CloudEnvironment.... it lists a bunch of urls, where the Azure API's are...

**Comment by paulmey** [comment]
> Does this change require documentation?
> I read your commit message, and it does not appear to do justice to the amount of change in this PR.

Yes it does require doc changes. A bearer JWT is a JWT derived from a certificate, but with a chosen expiration time. You can use this code to cert-based authentication with AAD in such a way that if someone steals your JWT they only have time-limited access to Azure. I also added a cmd to generate bearer JWT from a certificate.

I'll work with @amydutta to get the code refactored on top of the MSI PR, which I had missed going in.

File: builder/azure/cmd/mkbearerjwt/main.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+func main() {
+	var (
+		certfile string
+		duration time.Duration
+		addX5c   bool
+		audience string
+		clientID string
+	)
+	flag.StringVar(&clientID, "clientId", "", "The ID of the client that is going to use this token (sub)")
```
[the remark this answers is not in the document]
Comment (paulmey) [reply builder/azure/cmd/mkbearerjwt/main.go]: No, this is the `sub` claim (subject) in the JWT token. We should probably be a bit more verbose here... :wink:

File: builder/azure/cmd/mkbearerjwt/main.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+	flag.DurationVar(&duration, "validFor", time.Hour, "Validity duration")
```
[the remark this answers is not in the document]
Comment (paulmey) [reply builder/azure/cmd/mkbearerjwt/main.go]: It is not in hours, instead a [Go duration string](https://golang.org/pkg/time/#ParseDuration). the time.Hour is the default.

**Comment by paulmey** [comment]
superseeded by #7189

**Pull request closed without merging**